enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/PropertyExistenceEnforcer.java

```diff
@@ -42,191 +49,406 @@
+        {
+            update( mandatoryNodePropertiesByLabel, constraint.getLabelId(), constraint.getPropertyIds() );
+        }
+        for ( RelationTypeSchemaDescriptor constraint : rels )
         {
-            throw new NodePropertyExistenceException( descriptor, ConstraintValidationException.Phase.VALIDATION, nodeId );
+            update( mandaRelationshipPropertiesByType, constraint.getRelTypeId(), constraint.getPropertyIds() );
         }
     }
 
-    private Cursor<NodeItem> nodeCursor( long id )
+    private void validateNodeProperties( long id, PrimitiveIntSet labelIds, PrimitiveIntSet propertyKeyIds )
+            throws NodePropertyExistenceException
     {
-        Cursor<NodeItem> cursor = storeStatement().acquireSingleNodeCursor( id );
-        return txState.augmentSingleNodeCursor( cursor, id );
+        if ( labelIds.size() > mandatoryNodePropertiesByLabel.size() )
+        {
+            for ( PrimitiveIntIterator labels = mandatoryNodePropertiesByLabel.iterator(); labels.hasNext(); )
+            {
+                int label = labels.next();
+                if ( labelIds.contains( label ) )
+                {
+                    validateNodeProperties( id, label, mandatoryNodePropertiesByLabel.get( label ), propertyKeyIds );
+                }
+            }
+        }
+        else
+        {
+            for ( PrimitiveIntIterator labels = labelIds.iterator(); labels.hasNext(); )
+            {
+                int label = labels.next();
+                int[] keys = mandatoryNodePropertiesByLabel.get( label );
+                if ( keys != null )
+                {
+                    validateNodeProperties( id, label, keys, propertyKeyIds );
+                }
+            }
+        }
     }
 
-    private Cursor<PropertyItem> properties( NodeItem node )
+    private void validateNodeProperties( long id, int label, int[] requiredKeys, PrimitiveIntSet propertyKeyIds )
+            throws NodePropertyExistenceException
     {
-        Lock lock = node.lock();
-        Cursor<PropertyItem> cursor = storeStatement().acquirePropertyCursor( node.nextPropertyId(), lock );
-        return txState.augmentPropertyCursor( cursor, txState.getNodeState( node.id() ) );
+        for ( int key : requiredKeys )
+        {
+            if ( !propertyKeyIds.contains( key ) )
+            {
+                failNode( id, label, key );
+            }
+        }
     }
 
-    private StorageStatement storeStatement()
+    private void failNode( long id, int label, int propertyKey )
+            throws NodePropertyExistenceException
     {
-        return storageStatement == null ? storageStatement = storeLayer.newStatement() : storageStatement;
+        for ( LabelSchemaDescriptor constraint : nodeConstraints )
+        {
+            if ( constraint.getLabelId() == label && contains( constraint.getPropertyIds(), propertyKey ) )
+            {
+                throw new NodePropertyExistenceException( constraint, VALIDATION, id );
```

Comment:
Yes, I thought about that too, and thought that we could implement it in a way that we report _all_ violations. That seemed like a different change though, so I didn't do it at this stage.